### PR TITLE
[ENG-731] Check for minimal Python version

### PR DESCRIPTION
**Pull request by tebeka**
Checks for minimal Python version when creating a new Python runtime.

### CODE REVIEW

File: runtimes/pythonrt/py.go [hunk 1]
```diff
@@ -36,9 +37,40 @@ func createTar(fs fs.FS) ([]byte, error) {
...
+	Minor int
+}
+
 type exeInfo struct {
 	Exe     string
-	Version string
+	Version Version
+}
+
+func parsePyVersion(s string) (major, minor int, err error) {
```
Comment (ig-ra) [inline runtimes/pythonrt/py.go:50]: it feels a bit overcomplicated to me..
maybe just scanf?

File: runtimes/pythonrt/py.go [hunk 2]
```diff
@@ -56,7 +88,19 @@ func pyExeInfo(ctx context.Context) (exeInfo, error) {
 	}
 
 	version := strings.TrimSpace(buf.String())
-	return exeInfo{Exe: exePath, Version: version}, nil
+	major, minor, err := parsePyVersion(version)
+	if err != nil {
+		return exeInfo{}, fmt.Errorf("can't parse version %q: %w", version, err)
```
Comment (ig-ra) [inline runtimes/pythonrt/py.go]: "failed to parse python version" ? just to ensure that error is clear to user and related to python

File: runtimes/pythonrt/py_test.go [hunk 3]
```diff
@@ -78,3 +78,29 @@ func processEnv(t *testing.T, pid int) map[string]string {
...
+var pyVersionCases = []struct {
+	version string
+	major   int
+	minor   int
+	err     bool
+}{
+	{"Python 3.12.2", 3, 12, false},
+	{"Python 2.7.18", 2, 7, false},
+	{"Python 3", 0, 0, true},
+	{"", 0, 0, true},
```
Comment (ig-ra) [inline runtimes/pythonrt/py_test.go:93]: 3 more tests?
- "Python" , e.g. starts with right string, but there is no version at all
- "Python X.X.X more stuff"
- "meow X.X.X"

File: runtimes/pythonrt/pythonrt.go [hunk 4]
```diff
@@ -39,6 +39,19 @@ type pySvc struct {
 	comm      *Comm
 }
 
+var minPyVersion = Version{
```
Comment (ig-ra) [inline runtimes/pythonrt/pythonrt.go:42]: you already defined .python-version file. Could we ensure that this file is the single source of truth for min version?

File: runtimes/pythonrt/pythonrt.go [hunk 5]
```diff
@@ -53,7 +66,11 @@ func New() (sdkservices.Runtime, error) {
 		return nil, fmt.Errorf("python info: %w", err)
 	}
 
-	log.Info("system python info", zap.String("exe", info.Exe), zap.String("version", info.Version))
+	log.Info("system python info", zap.String("exe", info.Exe), zap.Any("version", info.Version))
+	if !isGoodVersion(info.Version) {
+		return nil, fmt.Errorf("python version %d.%d is too old, min is %d.%d", info.Version.Major, info.Version.Minor, minPyVersion.Major, minPyVersion.Minor)
```
Comment (ig-ra) [inline runtimes/pythonrt/pythonrt.go]: "python version required >= X.x. Found: Y.y"

File: runtimes/pythonrt/pythonrt_test.go [hunk 6]
```diff
@@ -141,3 +143,47 @@ func Test_pySvc_Run(t *testing.T) {
...
+	ok      bool
+}{
+	{Version{2, 6}, false},
+	{Version{3, minPyVersion.Minor - 1}, false},
+	{Version{3, minPyVersion.Minor}, true},
+	{Version{3, minPyVersion.Minor + 1}, true},
+}
+
+func Test_isGoodVersion(t *testing.T) {
+	for _, tc := range isGoodVersionCasess {
```
Comment (ig-ra) [inline runtimes/pythonrt/pythonrt_test.go:158]: - it's the same file/same dir in which you are adding versions. Consider to create it once + chmod before for loop?
- how this file is deleted?

**Review by ig-ra — CHANGES_REQUESTED** [review]

File: runtimes/pythonrt/py.go (hunk 1 above)
Comment (itayd) [reply runtimes/pythonrt/py.go:50]: or a regex?

Comment (tebeka) [reply runtimes/pythonrt/py.go:50]: Used scanf.

File: runtimes/pythonrt/py.go (hunk 2 above)
Comment (tebeka) [reply runtimes/pythonrt/py.go]: Done.

File: runtimes/pythonrt/py_test.go (hunk 3 above)
Comment (tebeka) [reply runtimes/pythonrt/py_test.go:93]: Done 1 and 3. Some Pythons can have more info, for example:
```
$ pypy3 --version 
Python 3.10.13 (fc59e61cfbff, Jan 17 2024, 05:35:45)
[PyPy 7.3.15 with GCC 13.2.1 20230801]
```

File: runtimes/pythonrt/pythonrt.go (hunk 4 above)
Comment (tebeka) [reply runtimes/pythonrt/pythonrt.go:42]: Using the `.python-version` file seems to magical. Also it's used for the current Python version for development which is bigger than the minimal one.

File: runtimes/pythonrt/pythonrt.go (hunk 5 above)
Comment (tebeka) [reply runtimes/pythonrt/pythonrt.go]: Done

File: runtimes/pythonrt/pythonrt_test.go (hunk 6 above)
Comment (tebeka) [reply runtimes/pythonrt/pythonrt_test.go:158]: Is this comment about `TestNewBadVersionBelow`? There's no file in this test.

`go test` deletes `t.TempDir()` after the tests.

File: runtimes/pythonrt/py.go [hunk 7]
```diff
@@ -36,9 +36,31 @@ func createTar(fs fs.FS) ([]byte, error) {
...
-	Version string
+	Exe           string
+	VersionString string
+	Version       Version
+}
+
+func parsePyVersion(s string) (major, minor int, err error) {
+	// Python 3.12.2
+	const prefix = "Python "
+	if len(s) < len(prefix) {
```
Comment (ig-ra) [inline runtimes/pythonrt/py.go]: I think there is no additional value in check length.
Maybe just move Python as %s, e.g. "%s %d.%d" to scanf?

File: runtimes/pythonrt/pythonrt.go (hunk 4 above)
Comment (ig-ra) [reply runtimes/pythonrt/pythonrt.go:42]: @tebeka unresolved back. let's talk?

File: runtimes/pythonrt/py.go (hunk 7 above)
Comment (tebeka) [reply runtimes/pythonrt/py.go]: Changed to `strings.HasPrefix`

**Review by ig-ra — APPROVED** [review]

File: runtimes/pythonrt/pythonrt.go (hunk 4 above)
Comment (tebeka) [reply runtimes/pythonrt/pythonrt.go:42]: Removed `.python-version`

**Pull request merged**